### PR TITLE
Various rustc_codegen_ssa cleanups

### PR DESCRIPTION
**Pull request by bjorn3**

### CODE REVIEW

File: compiler/rustc_codegen_gcc/src/common.rs [hunk 1]
```diff
@@ -28,6 +28,19 @@ impl<'gcc, 'tcx> CodegenCx<'gcc, 'tcx> {
         global
         // TODO(antoyo): set linkage.
     }
+
+    pub fn const_bitcast(&self, value: RValue<'gcc>, typ: Type<'gcc>) -> RValue<'gcc> {
+        if value.get_type() == self.bool_type.make_pointer() {
+            if let Some(pointee) = typ.get_pointee() {
+                if pointee.dyncast_vector().is_some() {
+                    panic!()
+                }
```
Comment (matthiaskrgr) [inline compiler/rustc_codegen_gcc/src/common.rs:37]: this could be an `assert!()`?

Comment (bjorn3) [reply compiler/rustc_codegen_gcc/src/common.rs:37]: This is just moving some code around.

**Comment by wesleywiser** [comment]
r? rust-lang/compiler

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
r? compiler-errors @bors r+

**Comment by tgross35** [comment]
Any chance this could have failed the rollup with https://github.com/rust-lang/rust/pull/127066#issuecomment-2196000966? That same error happened in another rollup too. Seems unlikely, but nothing else seems more likely

**Comment by bjorn3** [comment]
The failing test seems to be

```rust
// foo.rs

// This is kept because of `#[used]`:
#[used]
static FOO: u32 = 0;

// This is removable because it is unused:
#[allow(dead_code)]
static BAR: u32 = 0;

// This is kept because it is publicly reachable:
pub static BAZ: u32 = 0;

// This is kept because it is referenced by a public, reachable function:
static QUUX: u32 = 0;

pub fn quux() -> &'static u32 {
    &QUUX
}

// This is removable because it is referenced by a private, unused (dead) function:
static CORGE: u32 = 0;

#[allow(dead_code)]
fn corge() -> &'static u32 {
    &CORGE
}
```

Maybe https://github.com/rust-lang/rust/issues/127052 is related given that this uses `#[used]`?

**Pull request merged**